### DISCUSSION TITLE
LLGL and Bindless Resources

### DISCUSSION THREAD

**Discussion by gdianaty** [LukasBanana/LLGL#108] [Ideas]
Hello 👋

Recently, in my toy renderer that uses LLGL I was attempting to get up and running with bindless resources. As of Shader Model 6.6, HLSL is able to access dynamic resources (seen [here](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/HLSL_SM_6_6_DynamicResources.md)). There are also some really interesting conversations happening around bringing equivalent intrinsics for [DXC's Vulkan support](https://github.com/microsoft/DirectXShaderCompiler/issues/4255).

I do know that LLGL treats OpenGL as the minspec, and as I understand it [something like it is very possible in OpenGL](https://www.khronos.org/opengl/wiki/Bindless_Texture). Furthermore, Apple's Metal implements bindless rendering using [Argument Buffers](https://developer.apple.com/videos/play/wwdc2021/10286/). Overall this functionality seems like it's nearly universally available across APIs.

The question that stands out in my mind is this: would permitting the above functionality be possible? What (other than the obvious trivial root signature changes) would need to happen within LLGL? Is the described functionality even within the scope of LLGL?

Thank you.

**Comment by LukasBanana** [comment]
> I do know that LLGL treats OpenGL as the minspec

Yeah, I try to keep the GL backend compatible down to GL 2.x just to be able to run a simple test even on my VMs which don't have the best support for HW acceleration.

> would permitting the above functionality be possible?

I'm definitely interested in bringing more modern API features to LLGL and not all of them have to be supported across all APIs, since a 100% compatibility across all renderers is simply not feasible, Vulkan and Metal for instance don't support arbitrary sampler border colors while D3D12 only restricts them to static samplers, but those are rarely needed features to begin with.
Anything that is limited to a certain set of renderers should simply have their compatibility set mentioned in the documentation (e.g. as in [`borderColor`](https://github.com/LukasBanana/LLGL/blob/6477416a362f2eab6cc76147ac70b5385444e6c9/include/LLGL/SamplerFlags.h#L163)) and/or a respective field in the `RendererFeatures` struct.

> What (other than the obvious trivial root signature changes) would need to happen within LLGL?

There is at least one thing I want to fix first before I start adding new features and that is getting rid of `ResetResourceSlots`. I started with another approach but haven't found the time yet to finish it. It's been on my [TODO](https://github.com/LukasBanana/LLGL/issues/43) list for a long time now and it's bothering me that the D3D11 binding validation is giving me a hard time to find an efficient way to keep track of resource bindings to avoid the necessity for this function. No other backend needs this function and I dislike the concept of manually "unbinding" resources in an interface that allows a heap of resources to be bound at once. On top of that, `SetResource` uses mapped binding indices, i.e. the actual binding slot is defined in the pipeline layout while `ResetResourceSlots` uses direct binding slots - not a great design, but these things grew over time :)
Other than that, I assume the `ResourceHeap` and `PipelineLayout` interfaces are just the right place to support bindless resources.

> Is the described functionality even within the scope of LLGL?

Absolutely, and resource binding is probably one of the most important parts of a rendering abstraction layer. Other features like conditional rendering for instance are obsolete at this point but for now I just keep it in to support older hardware as well.

**Comment by gdianaty** [comment]
It's good to hear that there's interest in adding bindless resource management to LLGL! As I said I tried it but realized I was hacking around a little too much to be comfortable with charging forward, lol. I'm eager to help work out implementation specifics for the APIs, and how they should behave. I'd also like to volunteer but I'm admittedly still learning the newer DirectX APIs (I've been stuck on Source Engine for a long time, and it's D3D9).

I will admit that the lack of bindless resources has turned into somewhat of a blocker in my project. However, we can continue to charge forward in other places and use the existing binding model where needed.

I will note that on DX12, bindless resources are a first-class feature now, as where in Vulkan it's an extension. However, I find DX12's descriptor heap system to be exceedingly complex to use at the frontend. I think there's a lovely API to be had if we follow the overall grammar of LLGL. Admittedly, I know even less about the Vulkan API for bindless resources but I can certainly go find out!

**Comment by gdianaty** [comment]
Hello again @LukasBanana!

I was wondering if you had a particular timeline for this idea in mind, now that you've seemingly retired `ResetResourceSlots`. Are you still waiting to homogenize `CommandBuffer::SetResource` and the binding indices for resource heaps?

Trying to decide if I should implement a workaround for now. There is nothing more permanent than a temporary solution, lol.

**Reply by LukasBanana** [reply]
The short answer is, I do not have a timeline in mind as I don't want to rush out new features that I also haven't wrapped my head around yet.

Having said that, having more options for resource binding in LLGL is definitely a priority for me over other features like ray tracing or mesh shaders. Right now, LLGL only supports a single `ResourceHeap` per PSO layout, so that would probably have to be refactored as well. But perhaps there's an elegant solution that requires very little modification of the public interface that allows to use the bindless declarations but I have to look more into the new [HLSL Dynamic Resources](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_DynamicResources.html) documentation first.

Any proposals of how the interface could look like for bindless resources is appreciated.
Ideally, this would only require a single new flag in the `ResourceHeapDescriptor` or `PipelineLayoutDescriptor`.

**Reply by LukasBanana** [reply]
@gdianaty  I started with an experimental branch [`dev-bindless`](https://github.com/LukasBanana/LLGL/tree/dev-bindless) to explore this feature in the D3D12 backend:
[dev-bindless/tests/Testbed/UnitTests/TestBindless.cpp](https://github.com/LukasBanana/LLGL/blob/dev-bindless/tests/Testbed/UnitTests/TestBindless.cpp)
Please note that this branch will have a volatile interface as I'm trying to figure out what interface works with D3D12, Vulkan, and Metal.

**Reply by gdianaty** [reply]
Very exciting stuff! I'll take a deeper look soon, but at first blush the HLSL looks exactly the way I hoped it would.

**Reply by gdianaty** [reply]
@LukasBanana I gave it a go on my engine. Here's what's working:
- Mixing old-school `CommandBuffer::SetResource` and using bindless with `CommandBuffer::SetResourceHeap`. The API for the task is a bit odd, but certainly serviceable.
- Multiple resources in a given bindless heap (e.g. a diffuse texture and normal texture, and their respective samplers).

Here's what is **not working**:
- Multiple calls of `SetResourceHeap` in the same command buffer. From what I can tell, the first set of CBV_SRV_UAV descriptors that are submitted via `SetResourceHeap` is the only one that works correctly. Subsequent calls within the same command buffer are ignored. This results to the mildly amusing image of the Sponza scene, but with all surfaces using the leaf texture. Image attached. I'd like to stress that this is specific to a given command buffer, and does not apply, for example, to the entire LLGL::RenderSystem.

![image](https://github.com/LukasBanana/LLGL/assets/9407540/147908de-2b53-4435-b1ec-1a6b7b7da200)

**Reply by LukasBanana** [reply]
Thanks for the feedback. I can immediately tell why this happens. LLGL has to maintain a pool of descriptors that are successively placed after each other every time you bind a new PSO. This is because `SetDescriptorHeaps` in D3D12 can incur a pipeline flush (see D3D12 [Remarks](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12graphicscommandlist-setdescriptorheaps#remarks)) so the common approach is to bind a global descriptor heap once (or very rarely). That global heap is shader-visible and the user defined resource heaps are all non-shader-visible. When you call `SetResourceHeap`, the D3D12 backend copies all descriptors from the non-shader-visible heap into the shader-visible one with an incremented offset. This offset is not accounted for when you use direct indexing, so you would have to access that offset in your shaders when you access the global objects `ResourceDescriptorHeap` and `SamplerDescriptorHeap`. I will add this to the Bindless test.

**Reply by LukasBanana** [reply]
With a bindless resource heap, I would assume you don't have to call `SetResourceHeap` very often, since you have most of your resources in that one heap. @gdianaty how to you set up your frame that you have to call that function several times even with bindless? I am trying to come up with an idea how to include the internal offset into the API, since this is very D3D12 specific at the moment. I think in one way or another, you would have to add the offset into your shaders when you access the descriptor heaps.

**Reply by gdianaty** [reply]
In my renderer, there is a `LLGL::ResourceHeap` for every unique material in the given scene, and indices within that resource heap are placed into a conventionally bound material configuration buffer. This is why I have a multiplicity of them, and why I am mixing both bindless and normal binding.

Here's a snippet from my command buffer encoder for some context.

```cpp
	pCmdBuf->Begin();
	pCmdBuf->PushDebugGroup("Scene");
	{
		// Encode dynamic renderables.
		FOR_EACH_VEC(m_vecDynamicRenderables, nCurRenderable)
		{
			IRenderableDynamic* pRenderable = m_vecDynamicRenderables[nCurRenderable];

			CModel* pModel = static_cast<CModel*>(pRenderable->GetModel());

			FOR_RANGE(pModel->GetNumParts(), nCurPart)
			{
				// Get the current part that we're working on.
				const CMesh& curPart = pModel->GetPartEx(nCurPart);
				// Get the current material.
				const CMaterial* pMaterial = curPart.GetMaterialEx();

				// Get corresponding PSO for the current viewport.
				LLGL::PipelineState* pPSO = pViewport->GetPipleineStateForShader(pMaterial->GetShader());
				Assert(pPSO);

				// Set vertex and index buffers, business as usual.
				pCmdBuf->SetVertexBuffer(*curPart.GetVertexBuffer());
				pCmdBuf->SetIndexBuffer(*curPart.GetIndexBuffer());

				// The PSO is set, the resource heap is bound, and all of the
				// non-heap binding points are set.
				pCmdBuf->SetPipelineState(*pPSO);
				pCmdBuf->SetResourceHeap(*pMaterial->GetResourceHeap());
				pCmdBuf->SetResource(BINDINGPOS_GLOBALS, *pMaterial->GetGlobalsBuffer());
				pCmdBuf->SetResource(BINDINGPOS_VIEWPORT_CONFIG, *pViewport->GetViewportConfigBuf());
				pCmdBuf->SetResource(BINDINGPOS_SCENE_CONFIG, *m_pSceneConfigBuf);
				
				// Do the thing.
				pCmdBuf->DrawIndexed(curPart.GetIndexCount(), 0);
			}
		}
	}
	pCmdBuf->PopDebugGroup();
	pCmdBuf->End();
```

Let me know if you need any further context from me.

**Reply by LukasBanana** [reply]
The way your render pass is set up right now, you would have to increment two offsets for your descriptor heaps after every draw call by the number of SRV/UAV/CBV and Samplers in your pipeline layout and pass that into your shader. Such small constants that change every frame are generally a good candidate for uniforms, since the D3D12 backend makes use of the fast root parameter constants.
This is of course brittle, but since this is still experimental, you could increment the two offsets like this:
```cpp
std::uint32_t heapOffsets[2] = { 0, 0 }; // One for ResourceDescriptorHeap and one for SamplerDescriptorHeap
...
pCmdBuf->SetPipelineState(*pPSO);
pCmdBuf->SetResourceHeap(*pMaterial->GetResourceHeap());
pCmdBuf->SetResource(BINDINGPOS_GLOBALS, *pMaterial->GetGlobalsBuffer());
pCmdBuf->SetResource(BINDINGPOS_VIEWPORT_CONFIG, *pViewport->GetViewportConfigBuf());
pCmdBuf->SetResource(BINDINGPOS_SCENE_CONFIG, *m_pSceneConfigBuf);

// Set descriptor heap offsets
pCmdBuf->SetUniforms(0, heapOffsets, sizeof(heapOffsets));

// Do the thing.
pCmdBuf->DrawIndexed(curPart.GetIndexCount(), 0);

// Increment offsets
heapOffsets[0] += numResourceViewsInPSOLayout;
heapOffsets[1] += numSamplersInPSOLayout;
```
Your PSO layout could look like this:
```cpp
LLGL::PipelineLayout* pPSOLayout = pRenderer->CreatePipelineLayout(
    LLGL::Parse(
        "heap{ texture(...), sampler(...) }," // Pseudo syntax for your material resources and samplers
        "cbuffer(Globals@0):vert," // BINDINGPOS_GLOBALS
        "cbuffer(ViewportConfig@1):vert," // BINDINGPOS_VIEWPORT_CONFIG
        "cbuffer(SceneConfig@2):vert," // BINDINGPOS_SCENE_CONFIG
        "uint(ResourceHeapOffset)," // Offset into ResourceHeapOffset[]
        "uint(SamplerHeapOffset)," // Offset into SamplerDescriptorHeap[]
    )
);
```
The values for `numResourceViewsInPSOLayout` and `numSamplersInPSOLayout` depend on the total number of resource views and samplers (including the non-heap descriptors). This is the part that is brittle, because LLGL does not expose the internal offsets and this approach is basically tracking those values by reproducing their computation. Since this is D3D12 specific, I have yet to come up with a better solution. Your shader can then access the descriptor heaps like so:
```hlsl
uint ResourceHeapOffset;
uint SamplerHeapOffset;
float4 PSMain() : SV_Target {
    Texture2D colorMap = ResourceDescriptorHeap[ResourceHeapOffset + COLORMAP_DESCRIPTOR_INDEX];
    SamplerState colorMapSampler = SamplerDescriptorHeap[SamplerHeapOffset + SAMPLER_DESCRIPTOR_INDEX];
    ...
}
```
Please let me know if that is a functional workaround.

PS: I think you might gain better value from the descriptor heaps if you have all of your materials stored in just one `ResourceHeap`. This was also the idea before bindless in LLGL: Ideally, you have only one resource heap for each pipeline layout and then select the right one (i.e. `SetResourceHeap(*heap, descriptorSetIndexForMaterial)`) for each material. If you have one resource heap for each material, you basically have only one "set" in your heap and loose some of its benefits.

**Reply by gdianaty** [reply]
Thank you for the workaround.

What you specified is more or less what I anticipated having to do. However, as I was sharing the snippet with you I started to realize how grossly I am underutilizing the resource heaps with my current approach.

I'll reflect on how I'll build in a more sensible use of resource heaps, but I would also like to try the workaround for the time being. I'm planning on revisiting the way I build my scenes in the near future so a hacky work-around is welcome right now.

**Reply by gdianaty** [reply]
Hey @LukasBanana, I'm back with news.

I elected to go with your postscript suggestion and place all textures into one resource heap per pipeline layout, as it feels like it makes more sense in terms of API surface for other backends in the future (e.g. Vulkan). I also took into consideration your prior comments about speed advantages.

It worked swimmingly, and runs blazingly fast. 

However, a couple of things came up:

- The CBV/UAV/SRV Descriptor Heap in PIX is up to date during all my frame captures. However, the sampler descriptor heap takes exactly two encodings in order to be up to date with the last call of `->WriteResourceHeap`. I know this because I have to manually trigger a command buffer encoding two times in order for the sampler to update. I can try to come up with a sample or send a unlisted YouTube video to your e-mail if you'd like.
- Less pressingly, in both NVidia NSight Graphics and Microsoft PIX, the 0th item in the sampler descriptor heap is always displayed as invalid. To the best of my knowledge this should not work, yet, miraculously, all the GPUs I have tested have no problem reading `SamplerDescriptorHeap[0]`. This includes both NVidia and AMD chips. Admittedly I do not have a reference implementation of bindless renderer to cross-reference.

As per usual, let me know if I can help with anything.

**Reply by LukasBanana** [reply]
I'm happy to hear your refactor runs a lot faster now 👍 

As for your feedback:
1. The `WriteResourceHeap` issue sounds like a synchronization issue. Do you call this function during command encoding? Any function that is not part of the `CommandBuffer` interface is not guaranteed to take effect until that command buffer is submitted to the command queue, even an "immediate command buffer" does not guaranteed that. If you need to update resources dynamically you'll have to use the individual resource bindings. The resource heap is meant for resources that don't change during command encoding (i.e. within that heap) but you can quickly swap out sets within a resource heap. So if you want to use the resource heap for your materials and they all reference several resources (albedo, normal, roughness etc.) but only a single one is different in some of them (e.g. only the roughness map is different), you'll have to write duplicate entries into the resource heap because only the entire set can be swapped out.
2. If you can share some sample code and even a video or screenshot alongside that would be great.

**Reply by gdianaty** [reply]
I do not call the `WriteResourceHeap` function during encoding, but rather, before encoding ever starts. I'll look into making a reproducible sample but life has been keeping me busy, unfortunately.

Here's a video of the unexpected behavior in my engine's model viewer, and some screenshots from PIX highlighting my aforementioned weird behavior. Again to recap: the sampler descriptor heap appears to be out of sync by approximately one encoding, but it is not consistent.

If I call WriteResourceHeap twice before encoding (one texture and one sampler), the unexpected behavior occurs. Unusually, if I call it two times (two textures and two samplers), the unexpected behavior does not occur. Even more unusually if I call it several more times (e.g. the Sponza which has a total of 50 calls over all materials) the unexpected behavior occurs.

https://github.com/LukasBanana/LLGL/assets/9407540/6c0a5e66-dd12-4d15-ac5a-aa5a41f8aee8

![Untitled-1](https://github.com/LukasBanana/LLGL/assets/9407540/d12be5d9-c17b-4735-870b-a0e800b34f2f)

**Reply by LukasBanana** [reply]
> I do not call the WriteResourceHeap function during encoding, but rather, before encoding ever starts.

Ok good, that's how it's intended.

> I'll look into making a reproducible sample but life has been keeping me busy, unfortunately.

That would be fabulous of course :)

It does sounds like the out of sync issue could be a problem with the scheduling of the descriptor heaps internally. Have you tried using a command buffer with only a single native buffer, i.e. `CommandBufferDescriptor::numNativeBuffers=1`? If that fixes the issue, it could narrow down what causes the out of sync with sampler descriptors. This should only be a temporarily solution of course, because you wouldn't want to limit your command buffer to a single native command buffer unless it's encoded only once.

**Reply by gdianaty** [reply]
@LukasBanana 

I was able to try setting all my CommandBuffers (both the ones for the viewport that are per-buffer and the bundles used in my codebase) to use `numNativeBuffers = 1` and I had no immediate effect.

So no dice there, unfortunately.

**Reply by gdianaty** [reply]
@LukasBanana 

Hello! It's been a good while hasn't it :) lol

I'll be honest, I've overhauled my engine's resource heap system and made one very small (yet seemingly very large) change:

Before I was adding the texture and sampler using an initializer list call to `WriteResourceHeap`, something like this:

```
pLLGL->WriteResourceHeap(*pResourceHeap, nResourceHeapIdx, { pTexture, pSampler });
```

And then in the GPU, I would simply access `ResourceDescriptorHeap` and `SamplerDescriptorHeap` with the same value, `nResourceHeapIdx`.

However, now I perform them as two resource heap writes as a consequence of some internal API changes:

```
pLLGL->WriteResourceHeap(*pResourceHeap, nResourceHeapIdx, { pTexture });
// and then, later on in another part of the code:
pLLGL->WriteResourceHeap(*pResourceHeap, nResourceHeapIdx, { pSampler });
```

And it works just fine. Not really sure why this is the case but it is. I know you mentioned wanting to reimagine the API surface for WriteResourceHeap for bindless, especially for d3d12. Frankly, at the moment the current API for interacting with d3d12's distinctly separate descriptor heaps is less than ideal. As of right now, I am maintaining an internal cpu-side map of heap indices to resource pointers. This was ideal for my workflow and I'm happy to share more details if you think it could be valuable. I would love to see this feature merged into the main branch of LLGL, but I'm not sure what your plan would be for supporting this on Vulkan as well. I'm interested to hear your thoughts on that.

Broadly, though, bindless has been working great for our use case for the last several months and continues to be perfectly serviceable. Recently I implemented bindless StructuredBuffer vertex data streams -- super cool!

**Reply by LukasBanana** [reply]
I faintly remember that work and I haven't synced that branch in a while, I bet there will be quite a few merge conflicts, but I certainly wouldn't want this to go to waste.
I think this worked because the descriptor indices were mapped differently between the Sampler and Non-Sampler internal heaps. If that was depending on the input resource, that would explain why it behaves this way even though you use the same input index. That's not by design, however.
I wouldn't mind giving this another try now that I worked a bit more on and with the D3D12 backend.
Would love to see some of your work and how you use LLGL :)

**Reply by gdianaty** [reply]
> I think this worked because the descriptor indices were mapped differently

Yes I seem to vaguely remember reading the underlying code that's responsible for the behavior and coming up with this workaround. Again it would be awesome to refine the API a little and get it merged into main, but running on a stable but stale branch is not a show stopper by any means.

> Would love to see some of your work and how you use LLGL :)

My username on DIscord is `gmdianaty` and I'd love to show you some of what we're doing with LLGL but I'm not ready to share it with the world at large just yet :)
